File: src/main/java/fi/quanfoxes/assembler/VariableValue.java
package fi.quanfoxes.assembler;

import fi.quanfoxes.parser.Variable;

public class VariableValue extends Value {
    private Variable variable;

    public VariableValue(Register register, Variable variable) {
        super(register, Size.get(variable.getType().getSize()), ValueType.VARIABLE, true, false, true);
        this.variable = variable;
    }

    private VariableValue(VariableValue value) {
        super(value);
        this.variable = value.variable;
    }

    @Override
    public Value clone(Register register) {
        VariableValue clone = new VariableValue(this);
        register.attach(clone);
        return clone;
    }

    public Variable getVariable() {
        return variable;
    }
}

File: Zigzag/Assembler/VariableValue.cs
public class VariableValue : Value
{
	public Variable Variable { get; private set; }

	public VariableValue(Register register, Variable variable) : base(register, Size.Get(variable.Type.Size), ValueType.VARIABLE, true, false, true)
	{
		Metadata = variable;
		Variable = variable;
	}

	private VariableValue(VariableValue value) : base(value)
	{
		Variable = value.Variable;
	}

	public override Value Clone(Register register)
	{
		var clone = new VariableValue(this);
		register.Attach(clone);
		return clone;
	}
}